### PR TITLE
8297519: Improve expressions and modernise code in PKCS

### PR DESCRIPTION
**Pull request by minborg**
During the work of another PR (https://github.com/openjdk/jdk/pull/11260), several improvement areas were identified. These are now adressed in this separate PR proposing the use of more modern Java constructs as well as simplifying a large number of logical expressions that were previously non-normative. 

This branch has been tested and passed tier1-4 tests.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8297519](https://bugs.openjdk.org/browse/JDK-8297519): Improve expressions and modernise code in PKCS

### Reviewers
 * [Weijun Wang](https://openjdk.org/census#weijun) (@wangweij - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11348/head:pull/11348` \
`$ git checkout pull/11348`

Update a local copy of the PR: \
`$ git checkout pull/11348` \
`$ git pull https://git.openjdk.org/jdk pull/11348/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11348`

View PR using the GUI difftool: \
`$ git pr show -t 11348`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11348.diff">https://git.openjdk.org/jdk/pull/11348.diff</a>

</details>

### CODE REVIEW

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11AEADCipher.java [hunk 1]
```diff
@@ -311,18 +310,10 @@ protected void engineInit(int opmode, Key key, AlgorithmParameters params,
...
-                            params.getParameterSpec(GCMParameterSpec.class);
-                        break;
-                    case CHACHA20_POLY1305:
-                        paramSpec =
-                            params.getParameterSpec(IvParameterSpec.class);
-                        break;
-                    default:
-                        throw new AssertionError("Unsupported type " + type);
-                }
+                paramSpec = switch (type) {
```
Comment (minborg) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11AEADCipher.java:313]: This switch is exhaustive.

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/Secmod.java [hunk 2]
```diff
@@ -442,26 +447,13 @@ public static final class Module {
...
-            case FIPS:
-                config = String.format(TEMPLATE_FIPS, libraryName);
-                break;
-            case TRUSTANCHOR:
-                config = String.format(TEMPLATE_TRUSTANCHOR, libraryName);
-                break;
-            default:
-                throw new RuntimeException("Unknown module type: " + type);
-            }
+            config = switch (type) {
```
Comment (minborg) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/Secmod.java:450]: Exhaustive

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/Secmod.java [hunk 3]
```diff
@@ -353,35 +353,40 @@ public Module getModule(ModuleType type) {
...
-        + "name = \"NSS SoftToken KeyStore\"\n"
-        + "slotListIndex = 1\n"
-        + "nssUseSecmodTrust = true\n";
-
-    static final String TEMPLATE_FIPS =
-        "library = %s\n"
-        + "name = \"NSS FIPS SoftToken\"\n"
-        + "slotListIndex = 0\n"
-        + "nssUseSecmodTrust = true\n";
+    static final String TEMPLATE_EXTERNAL =  """
```
Comment (turbanoff) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/Secmod.java:356]:
```suggestion
    static final String TEMPLATE_EXTERNAL = """
```

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/Config.java [hunk 4]
```diff
@@ -479,14 +479,11 @@ private void parse() throws IOException {
...
-                } else if (mode.equals("noDb")) {
-                    nssDbMode = Secmod.DbMode.NO_DB;
-                } else {
-                    throw excToken("nssDbMode must be one of readWrite, readOnly, and noDb:");
+                switch (mode) {
+                    case "readWrite" -> nssDbMode = Secmod.DbMode.READ_WRITE;
+                    case "readOnly" -> nssDbMode = Secmod.DbMode.READ_ONLY;
+                    case "noDb" -> nssDbMode = Secmod.DbMode.NO_DB;
+                    default -> throw excToken("nssDbMode must be one of readWrite, readOnly, and noDb:");
                 }
```
Comment (ExE-Boss) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/Config.java:487]:
```suggestion
                nssDbMode = switch (mode) {
                    case "readWrite" -> Secmod.DbMode.READ_WRITE;
                    case "readOnly" -> Secmod.DbMode.READ_ONLY;
                    case "noDb" -> Secmod.DbMode.NO_DB;
                    default -> throw excToken("nssDbMode must be one of readWrite, readOnly, and noDb:");
                };
```

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/Config.java [hunk 5]
```diff
@@ -1044,14 +1034,11 @@ private void parseHandleStartupErrors(String keyword) throws IOException {
...
-        } else if (val.equals("halt")) {
-            handleStartupErrors = ERR_HALT;
-        } else {
-            throw excToken("Invalid value for handleStartupErrors:");
+        switch (val) {
+            case "ignoreAll" -> handleStartupErrors = ERR_IGNORE_ALL;
+            case "ignoreMissingLibrary" -> handleStartupErrors = ERR_IGNORE_LIB;
+            case "halt" -> handleStartupErrors = ERR_HALT;
+            default -> throw excToken("Invalid value for handleStartupErrors:");
         }
```
Comment (ExE-Boss) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/Config.java:1042]:
```suggestion
        handleStartupErrors = switch (val) {
            case "ignoreAll" -> ERR_IGNORE_ALL;
            case "ignoreMissingLibrary" -> ERR_IGNORE_LIB;
            case "halt" -> ERR_HALT;
            default -> throw excToken("Invalid value for handleStartupErrors:");
        };
```

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Cipher.java [hunk 6]
```diff
@@ -226,18 +226,17 @@ protected void engineSetMode(String mode) throws NoSuchAlgorithmException {
...
             }
-            result = MODE_CBC;
-        } else if (mode.equals("CTR")) {
-            result = MODE_CTR;
-        } else {
-            throw new NoSuchAlgorithmException("Unsupported mode " + mode);
+            case "CTR" -> result = MODE_CTR;
+            default -> throw new NoSuchAlgorithmException("Unsupported mode " + mode);
         }
         return result;
```
Comment (ExE-Boss) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Cipher.java:241]:
```suggestion
        return switch (mode) {
            case "ECB" -> MODE_ECB;
            case "CBC" -> {
                if (blockSize == 0) {
                    throw new NoSuchAlgorithmException
                            ("CBC mode not supported with stream ciphers");
                }
                yield MODE_CBC;
            }
            case "CTR" -> MODE_CTR;
            default -> throw new NoSuchAlgorithmException("Unsupported mode " + mode);
        };
```

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Cipher.java [hunk 7]
```diff
@@ -364,19 +363,13 @@ private void implInit(int opmode, Key key, byte[] iv,
...
-            default:
+            case Cipher.ENCRYPT_MODE -> encrypt = true;
+            case Cipher.DECRYPT_MODE -> encrypt = false;
+            case Cipher.WRAP_MODE, Cipher.UNWRAP_MODE -> throw new UnsupportedOperationException
+                    ("Unsupported mode: " + opmode);
+            default ->
                 // should never happen; checked by Cipher.init()
-                throw new AssertionError("Unknown mode: " + opmode);
+                    throw new AssertionError("Unknown mode: " + opmode);
         }
```
Comment (ExE-Boss) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Cipher.java:373]:
```suggestion
        encrypt = switch (opmode) {
            case Cipher.ENCRYPT_MODE -> true;
            case Cipher.DECRYPT_MODE -> false;
            case Cipher.WRAP_MODE, Cipher.UNWRAP_MODE -> throw new UnsupportedOperationException
                    ("Unsupported mode: " + opmode);
            default ->
                // should never happen; checked by Cipher.init()
                    throw new AssertionError("Unknown mode: " + opmode);
        };
```

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11RSACipher.java [hunk 8]
```diff
@@ -385,32 +369,25 @@ private int implDoFinal(byte[] out, int outOfs, int outLen)
...
-            case MODE_VERIFY:
-                n = p11.C_VerifyRecover
+                case MODE_VERIFY -> n = p11.C_VerifyRecover
                         (session.id(), buffer, 0, bufOfs, out, outOfs, outLen);
-                break;
-            default:
-                throw new ProviderException("internal error");
+                default -> throw new ProviderException("internal error");
             }
             return n;
```
Comment (ExE-Boss) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11RSACipher.java:392]:
```suggestion
            return switch (mode) {
                case MODE_ENCRYPT -> p11.C_Encrypt
                        (session.id(), 0, buffer, 0, bufOfs, 0, out, outOfs, outLen);
                case MODE_DECRYPT -> p11.C_Decrypt
                        (session.id(), 0, buffer, 0, bufOfs, 0, out, outOfs, outLen);
                case MODE_SIGN -> {
                    byte[] tmpBuffer = new byte[bufOfs];
                    System.arraycopy(buffer, 0, tmpBuffer, 0, bufOfs);
                    tmpBuffer = p11.C_Sign(session.id(), tmpBuffer);
                    if (tmpBuffer.length > outLen) {
                        throw new BadPaddingException(
                                "Output buffer (" + outLen + ") is too small to " +
                                        "hold the produced data (" + tmpBuffer.length + ")");
                    }
                    System.arraycopy(tmpBuffer, 0, out, outOfs, tmpBuffer.length);
                    yield tmpBuffer.length;
                }
                case MODE_VERIFY -> p11.C_VerifyRecover
                        (session.id(), buffer, 0, bufOfs, out, outOfs, outLen);
                default -> throw new ProviderException("internal error");
            };
```

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11KeyWrapCipher.java [hunk 9]
```diff
@@ -744,20 +740,20 @@ protected Key engineUnwrap(byte[] wrappedKey, String wrappedKeyAlgo,
...
-            default: // should never happen
-                throw new AssertionError();
-        };
+            }
+            case Cipher.PUBLIC_KEY -> throw new UnsupportedOperationException
+                    ("cannot unwrap public keys");
+            default -> // should never happen
+                    throw new AssertionError();
+        }
+        ;
```
Comment (ExE-Boss) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11KeyWrapCipher.java:756]: Empty statement:
```suggestion
```

**Review by ExE-Boss — COMMENTED** [review]
These can all be `switch` expressions:

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11AEADCipher.java [hunk 10]
```diff
@@ -373,18 +362,17 @@ private void implInit(int opmode, Key key, byte[] iv, int tagLen,
...
+                }
+                case CHACHA20_POLY1305 -> {
                     iv = new byte[type.defIvLen];
                     this.random.nextBytes(iv);
                     tagLen = type.defTagLen;
-                    break;
-                default:
-                    throw new AssertionError("Unsupported type " + type);
+                }
+                default -> throw new AssertionError("Unsupported type " + type);
```
[the remark this answers is not in the document]
Comment (minborg) [reply src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11AEADCipher.java:375]: This would enable total removal of the switch statement.

**Comment by RogerRiggs** [comment]
A suggestion on the bug title, include "PKCS" or something to identity what code is being modernized.

**Pull request closed without merging**